benchmark/torch/RL4LMs/rl4lms_utils/rollout_util.py

```diff
@@ -0,0 +1,239 @@
+#   Copyright (c) 2023 PaddlePaddle Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import torch
+import numpy as np
+
+from .kl_controller import KLController
+from parl.utils import logger
+from collections import OrderedDict
+from .data_wrapper import TransitionInfo
+
+
+def get_one_token_obs(obs, idx, space):
+    return OrderedDict([(k, obs[k][:, idx, :]) for k in space.spaces.keys()])
+
+
+def unpack_observations(obs_tensor, n_instructors):
+    """
+    Unpacks vectorized dict observations into separate dict observations
+    """
+    unpacked_obs = []
+    keys = obs_tensor.keys()
+    for instructor_ix in range(n_instructors):
+        obs_dict = {}
+        for key in keys:
+            obs_dict[key] = obs_tensor[key][instructor_ix].reshape(1, -1).cpu()
+        unpacked_obs.append(obs_dict)
+    return unpacked_obs
+
+
+class RolloutUtil:
+    def __init__(self, kl_args):
+        self._kl_controller = KLController(kl_args["coeff"], kl_args["target_kl"])
+
+    def collect_rollouts(self, agent, instructor_group, rollout_buffer):
+        # get tokenizer
+        tokenizer = instructor_group.tokenizer
+
+        # Switch to eval mode both training and testing
+        agent.eval_mode()
+
+        # reset rollout buffer and stats
+        rollout_buffer.reset()
+
+        # start the rollout process
+        rollout_info = {
+            "rollout_info/ep_rew": [],
+            "rollout_info/kl_div_mean": [],
+            "rollout_info/ep_lens": [],
+            "rollout_info/ep_kl_rew": [],
+            "rollout_info/log_prob": [],
+            "rollout_info/ref_log_prob": [],
+            "rollout_info/values": [],
+        }
+        num_timesteps = 0
+        while not rollout_buffer.full:
+            # start parallel episodes
+            current_obs = instructor_group.ask()
+
+            # generate sentences using the model
+            generation_inputs = agent.get_inputs_for_generation(current_obs)
```

Comment:
Suggestion: rename `get_inputs_for_generation` as `build_inputs`, and integrate it into `agent.predict()`, which means we can simply inference `gen_output` from `current_obs` by `agent.predict()`